### PR TITLE
Convert sdetools to APIv2

### PR DESCRIPTION
**Pull request by geoffwhittington**
Almost there...

- [x] Mock tests pass
- [x] Live tests pass
- [x] Change /api/v2/taskstatuses => /api/v2/task-statuses (merge request is open)
- [x] Squash commits
- [x] Merge in master
- [x] Test all syncs/imports manually and check for small details

### CODE REVIEW

**Comment by funkaoshi** [comment]
Good start.

File: sdetools/sdelib/sdeapi.py [hunk 1]
```diff
@@ -94,64 +91,43 @@ def get_projects(self, application, options={}, **filters):
...
-            return
-        if note_type:
-            end_point += '/%s' % (note_type)
-        args = {'task': task}
-        args.update(filters)
-        return self.call_api(end_point, args=args, call_headers=_encode_options(options))
+        return self.call_api('projects/%d/tasks/%s/notes/%d/' % (project_id, task_id, note_id), self.URLRequest.PUT, args=note)
+
+    def get_task_notes(self, project_id, task_id, options={}, **filters):
+        args =
```
Comment (geoffwhittington) [inline sdetools/sdelib/sdeapi.py]: What is the missing code here?

**Comment by geoffwhittington** [comment]
Seems like we have 5 test errors remaining. Almost there....

```
  File "/home/vagrant/sdetools/sdetools/analysis_integration/tests/base_integration_test.py", line 224, in test_import_findings_assert_failed_tasks
    result = self.integrator.import_findings()
  File "/home/vagrant/sdetools/sdetools/analysis_integration/base_integrator.py", line 575, in import_findings
    project_analysis_note_ref = ret['id']
KeyError: 'id'
```

```
    return wrapped(self, *args, **kwargs)
  File "/home/vagrant/sdetools/sdetools/sdelib/interactive_plugin.py", line 233, in get_taskstatuses
    return self.api.get_taskstatuses()
  File "/home/vagrant/sdetools/sdetools/sdelib/sdeapi.py", line 179, in get_taskstatuses
    return response['results']
KeyError: 'results'
```

```
    self.sde_plugin.update_task_status(task['id'], status)
  File "/home/vagrant/sdetools/sdetools/sdelib/interactive_plugin.py", line 12, in wrapper
    return wrapped(self, *args, **kwargs)
  File "/home/vagrant/sdetools/sdetools/sdelib/interactive_plugin.py", line 237, in update_task_status
    return self.api.update_task_status(self.prj_id, task_id, status)
  File "/home/vagrant/sdetools/sdetools/sdelib/sdeapi.py", line 172, in update_task_status
    return result['status']
KeyError: 'status'
```
```
  File "/home/vagrant/sdetools/sdetools/modules/sync_hp_alm/tests/hp_alm_plugin_test.py", line 69, in test_add_test_plan
    self.connector.alm_add_task(test_task)
  File "/home/vagrant/sdetools/sdetools/modules/sync_hp_alm/hp_alm_plugin.py", line 346, in alm_add_task
    alm_task = self._alm_add_test_plan(task_id, field_data, task)
  File "/home/vagrant/sdetools/sdetools/modules/sync_hp_alm/hp_alm_plugin.py", line 318, in _alm_add_test_plan
    req_ids = self.requirement_to_test_mapping.get(task['weakness']['id'])
KeyError: 'weakness'
```

```
Traceback (most recent call last):
  File "/home/vagrant/sdetools/sdetools/modules/sync_rational/tests/rational_plugin_test.py", line 50, in test_update_task_status_to_todo
    self.assertEqual(the_task['status'], 'TODO', 'Failed to update SDE task to TODO')
KeyError: 'status'
```

**Comment by geoffwhittington** [comment]
No idea why tests are failing on Travis - there must be some file differences between local and here

**Comment by geoffwhittington** [comment]
Mock tests are failing because we refer to 'description' in the code and the mocked responses use 'text'

File: sdetools/analysis_integration/tests/base_integration_test.py [hunk 2]
```diff
@@ -225,7 +225,7 @@ def test_import_findings_assert_failed_tasks(self):
 
         self.assertEqual(len(result.affected_tasks), 1, 'Expected 1 affected tasks, got %s' % len(result.affected_tasks))
         self.assertTrue('36' in result.affected_tasks, 'Expected to find task 36 in affected tasks')
-        self.check_analysis_note('36', 'failed', 'high', 1)
+        self.check_analysis_note('36', 'failed', 'low', 1)
```
Comment (funkaoshi) [inline sdetools/analysis_integration/tests/base_integration_test.py]: Why did this change?

File: sdetools/sdelib/sdeapi.py [hunk 3]
```diff
@@ -39,13 +39,8 @@ def post_conf_init(self):
...
-            self.set_auth_mode('session')
-            result = True
-        else:
-            #TODO: Find a better alternative ->
-            #In 'basic' mode, we make an extra call just to verify that credentials are correct
-            result = self.get_users(myself_only=True)
+        # Simple check
+        result = self.get_users(myself_only=True)
         self.connected = True
         return not not result
```
Comment (funkaoshi) [inline sdetools/sdelib/sdeapi.py]: Seems clearer to say `bool(result)` if the goal is to turn this to a boolean value or something.

File: sdetools/sdelib/sdeapi.py [hunk 4]
```diff
@@ -77,13 +72,15 @@ def get_applications(self, options={}, **filters):
...
+        response = self.call_api('applications/', args=filters,
                 call_headers=_encode_options(options))
-        return result['applications']
+        return response['results']
 
-    def create_application(self, name):
-        args = {'name': name}
-        return self.call_api('applications', self.URLRequest.POST, args=args)
+    def create_application(self, name, business_unit):
+        args = {'business_unit': business_unit,
```
Comment (funkaoshi) [inline sdetools/sdelib/sdeapi.py]: Normally we format this stuff:

```
args = {
    'business_unit': ...
    ...
}
```

File: sdetools/sdelib/testlib/sde_response_generator.py [hunk 5]
```diff
@@ -200,72 +239,55 @@ def call_task(self, target, flag, data, method):
...
-            else:
-                data['status'] = 'failed'
-                data['confidence'] = 'high'
         else:
-            self.raise_error('500', {'error': 'Missing parameter'})
+            self.raise_error('401')
 
-        self.generator_add_resource(note_type, resource_data=data)
+    def call_task_analysis_note(self, target, flag, data, method):
+        if not flag:
```
Comment (funkaoshi) [inline sdetools/sdelib/testlib/sde_response_generator.py]: You can unindent a big chunk of this method if you start it off with:

```
if flag:
    self.raise_error('401')

task_id = target.split('/')[-3].split('-')[1]
...
```

File: sdetools/analysis_integration/tests/base_integration_test.py (hunk 2 above)
Comment (jnlantos) [inline sdetools/analysis_integration/tests/base_integration_test.py]: I'm not sure why this was changed either. I've cleaned things up according to the rest of your comments though.

Comment (geoffwhittington) [inline sdetools/analysis_integration/tests/base_integration_test.py]: In the old response generator we always returned confidence as High for project analysis notes.  See the diff code for `add_analysis_session`.

In the new code we return back what was sent originally.

File: sdetools/modules/command_driver.py [hunk 6]
```diff
@@ -77,22 +72,37 @@ def filter_connections(self, connections):
...
         self.sde_connect()
 
         logger.info('Retrieving connection details')
 
-        try:
-            response = self.api.get_integration_connections()
-            connections = response['connections']
-        except APIError:
-            raise Error('Unable to retrieve connections from SD Elements')
+        connections = self.get_connections()
```
Comment (lmontopo) [inline sdetools/modules/command_driver.py]: Should this be in a try block, or no?

Comment (jnlantos) [inline sdetools/modules/command_driver.py]: The try happens in the get_connections() function instead. So no.

Comment (lmontopo) [inline sdetools/modules/command_driver.py]: OK cool :)

**Comment by lmontopo** [comment]
I think this looks good, but I should spend more time reviewing the second half of it tomorrow, as my brain was starting to get foggy by the end.

**Comment by geoffwhittington** [comment]
When was master last merged in? I've had to merge checkmarx and version one sync into master since this PR. I expect there will be a need to change the versionone code to address `task['status']['slug']` and `task['title']` but I don't anticipate much else.

**Comment by jnlantos** [comment]
Closing because squashed PR is here: https://github.com/sdelements/sdetools/pull/313

**Pull request closed without merging**